core/controllers/domain_objects_validator.py

```diff
@@ -178,3 +178,14 @@ def validate_aggregated_stats(aggregated_stats):
                 raise base.BaseHandler.InvalidInputException(
                     '%s not in state stats mapping of %s in aggregated '
                     'stats dict.' % (state_stats_property, state_name))
+
+
+def validate_params_dict(params):
+    """validates params dict
+
+    Args:
+        params: dict. Data that needs to be validated.
+    """
+    for key in params.keys():
+        if not isinstance(key, str):
+            raise Exception('Excepted str, received %s' % key)
```

Comment:
1. @Nik-09 here I have tried to add a check for `key` and I have also added a test in `core/controllers/reader_test.py` with non-string type data. In the test, I was expecting status code 400 but I am getting 200.
2.  [Here](https://github.com/oppia/oppia/blob/develop/core/storage/statistics/gae_models.py#L453) `params` is following `dict[str, str]` so, can we keep the rules for values?

Reply:
For point 1. -- You have to debug by adding print logs in every stage of the architecture, this is the general way you can solve a bug. Also I think you need to add backend tests in domain_objects_validator_test.py file.

For point 2 -- This is amibigious as in frontend params type is mentioned as Object and here in the backend it is mentioned as str (also with a note of confusion as given [here](https://github.com/oppia/oppia/blob/develop/core/storage/statistics/gae_models.py#L441))
As far I remember you reproduced e2e tests failures because  of type str but not sure.
@vojtechjelinek Can you suggest something for this ?

Reply:
@Nik-09 I have tried to write a test. Please have a look.